### PR TITLE
Add metadata considerations #63

### PR DESCRIPTION
**Pull request by claucece**

### CODE REVIEW

**Comment by claucece** [comment]
cc.: @alxdavids @chris-wood @dvorak42

File: draft-ietf-privacypass-architecture.md [hunk 1]
```diff
@@ -448,6 +448,51 @@ Metadata may also be arbitrarily long or bounded in length. The amount of
 permitted metadata may be determined by application or by the underlying
 cryptographic protocol.
 
+## Adding public metadata to the protocol
```
Comment (chris-wood) [inline draft-ietf-privacypass-architecture.md:451]: I'd recommend removing this entire section and folding it into the second paragraph above. Specifically, I might note that a timestamp-based epoch or the client's ASN could be used as transparent public metadata.

File: draft-ietf-privacypass-architecture.md [hunk 2]
```diff
@@ -448,6 +448,51 @@ Metadata may also be arbitrarily long or bounded in length. The amount of
...
+number will be increased by 1. This epoch number `n` will be cryptographically
+bounded (added as metadata) to the tokens by the server during the Issue phase,
+and stored by the client alongside with the tokens. During the redemption phase,
+the client will send this stored epoch (the number `n`) alongside the token. The
+server will check if their epoch is smaller than the client's sent one. If it
+is smaller, the token is expired. If it is the same, the token is valid. If it
+is bigger, the client might be trying to trick the server: this attack won't
+work as the Verification procedure executed by the server will fail.
+
+The anonymity set is still reduced depending on the key schedule of the server.
```
Comment (chris-wood) [inline draft-ietf-privacypass-architecture.md:493]: This is a useful paragraph to keep. Perhaps by appending it to the last paragraph in the section above?

File: draft-ietf-privacypass-protocol.md [hunk 3]
```diff
@@ -248,22 +248,31 @@ These requirements are covered in {{sec-reqs}}.
...
+generates a keypair by running `(pkS, skS) = KeyGen()`.
+The server and client also determine the public metadata that they will
+like to add, and the server is aware of the client metadata ahead of time.
+Metadata must be of the type `Metadata`, an opaque byte string of
+arbitrary length representing. This configuration must be available to all
+clients that interact with the server (for the purpose of engaging in a
+Privacy Pass exchange). We assume that the server has a public (and unique)
+identity that the client uses to retrieve this configuration.
+In {{draft-davidson-pp-architecture}} we discuss the kind of public
+metadata that is allowed.
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:260]:
```suggestion
```

File: draft-ietf-privacypass-protocol.md [hunk 4]
```diff
@@ -273,31 +282,31 @@ round sees the server generate a commitment. The second round sees the
 server issue a token to the client.
 
 ~~~
-  Client(pkS, m, info)                        Server(skS, pkS)
-  ------------------------------------------------------------
+  Client(pkS, m, info, cMetadata)                        Server(skS, pkS, sMetadata)
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:285]: Let's spell out these parameters? That is, `cMetadata` => `clientMetadata`.

File: draft-ietf-privacypass-protocol.md [hunk 5]
```diff
@@ -248,22 +248,31 @@ These requirements are covered in {{sec-reqs}}.
...
-mechanisms that the client can use to ensure that this public key is
-consistent across the entire ecosystem.
+time. The server and client also determine the public metadata that they will
+like to add, and the client is aware of the server metadata ahead of time.
+Metadata must be of the type `Metadata`, an opaque byte string of
+arbitrary length representing. The client picks a value `m` of tokens to
+receive during the issuance phase. In {{draft-davidson-pp-architecture}} we
+discuss mechanisms that the client can use to ensure that this public key
+is consistent across the entire ecosystem, and the kind of public metadata
+that is allowed.
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:275]:
```suggestion
is consistent across the entire ecosystem.
```

File: draft-ietf-privacypass-protocol.md (hunk 4 above)
Comment (chris-wood) [reply draft-ietf-privacypass-protocol.md:285]: Also, shouldn't both parties have client and server metadata as input?

File: draft-ietf-privacypass-protocol.md [hunk 6]
```diff
@@ -422,6 +431,15 @@ opaque PublicKey<1..2^16-1>
 opaque PrivateKey<1..2^16-1>
 ~~~
 
+### Public Metadata {#pp-public-metadata}
+
+We use the following type to describe the metadata that can be added
+by the server and/or client.
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:437]:
```suggestion
by the server or client.
```

File: draft-ietf-privacypass-protocol.md [hunk 7]
```diff
@@ -422,6 +431,15 @@ opaque PublicKey<1..2^16-1>
...
+
+~~~
+opaque Metadata<1..2^16-1>
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:440]: This means that metadata is length-prefixed, which isn't the case as described in text above. Maybe we just say that it's an arbitrary-length byte string and leave it at that? Then we could delete this section entirely. (See the proposed suggestions below!)

File: draft-ietf-privacypass-protocol.md [hunk 8]
```diff
@@ -597,6 +615,8 @@ Inputs:
 - `pkS`: A server `PublicKey`.
 - `skS`: A server `PrivateKey`.
 - `req`: An `IssuanceRequest` struct.
+- `sMetadata`: A optional server metadata `Metadata`.
+- `cMetadata`: A optional client metadata `Metadata`.
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:619]:
```suggestion
- `sMetadata`: A optional server metadata, an arbitrary-length byte string.
- `cMetadata`: A optional client metadata, an arbitrary-length byte string.
```

File: draft-ietf-privacypass-protocol.md [hunk 9]
```diff
@@ -616,6 +636,8 @@ Inputs:
 - `pkS`: An server `PublicKey`.
 - `input`: An `IssuanceInput` struct.
 - `resp`: An `IssuanceResponse` struct.
+- `sMetadata`: A optional server metadata `Metadata`.
+- `cMetadata`: A optional client metadata `Metadata`.
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:640]:
```suggestion
- `sMetadata`: A optional server metadata, an arbitrary-length byte string.
- `cMetadata`: A optional client metadata, an arbitrary-length byte string.
```

File: draft-ietf-privacypass-protocol.md [hunk 10]
```diff
@@ -638,6 +660,8 @@ Inputs:
 - `info`: An `opaque<1..2^16-1>` type corresponding to data that is
   linked to the redemption. See {{client-info}} for advice on how to
   construct this.
+- `sMetadata`: A optional server metadata `Metadata`.
+- `cMetadata`: A optional client metadata `Metadata`.
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:664]:
```suggestion
- `sMetadata`: A optional server metadata, an arbitrary-length byte string.
- `cMetadata`: A optional client metadata, an arbitrary-length byte string.
```

File: draft-ietf-privacypass-protocol.md [hunk 11]
```diff
@@ -653,6 +677,8 @@ Inputs:
 - `pkS`: An server `PublicKey`.
 - `skS`: An server `PrivateKey`.
 - `req`: A `RedemptionRequest` struct.
+- `sMetadata`: A optional server metadata `Metadata`.
+- `cMetadata`: A optional client metadata `Metadata`.
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:681]:
```suggestion
- `sMetadata`: A optional server metadata, an arbitrary-length byte string.
- `cMetadata`: A optional client metadata, an arbitrary-length byte string.
```

File: draft-ietf-privacypass-protocol.md [hunk 12]
```diff
@@ -248,22 +248,31 @@ These requirements are covered in {{sec-reqs}}.
...
-generates a keypair by running `(pkS, skS) = KeyGen()`. This
-configuration must be available to all clients that interact with the
-server (for the purpose of engaging in a Privacy Pass exchange). We
-assume that the server has a public (and unique) identity that the
-client uses to retrieve this configuration.
+generates a keypair by running `(pkS, skS) = KeyGen()`.
+The server and client also determine the public metadata that they will
+like to add, and the server is aware of the client metadata ahead of time.
+Metadata must be of the type `Metadata`, an opaque byte string of
+arbitrary length representing. This configuration must be available to all
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:255]:
```suggestion
The client and server also agree on public metadata values used during the protocol.
See {{client-setup}} for more information about this metadata. This configuration must be available to all
```

File: draft-ietf-privacypass-protocol.md [hunk 13]
```diff
@@ -248,22 +248,31 @@ These requirements are covered in {{sec-reqs}}.
...
 is a map of server identifiers (`server.id`) to arrays of stored tokens.
 We assume that the client knows the server public key `pkS` ahead of
-time. The client picks a value `m` of tokens to receive during the
-issuance phase. In {{draft-davidson-pp-architecture}} we discuss
-mechanisms that the client can use to ensure that this public key is
-consistent across the entire ecosystem.
+time. The server and client also determine the public metadata that they will
+like to add, and the client is aware of the server metadata ahead of time.
+Metadata must be of the type `Metadata`, an opaque byte string of
+arbitrary length representing. The client picks a value `m` of tokens to
```
Comment (chris-wood) [inline draft-ietf-privacypass-protocol.md:271]:
```suggestion
As above, the client and server also agree on public metadata values used during the protocol.
Each metadata value is an arbitrary-length byte string. The client picks an integer value `m > 0` of tokens to
```

**Review by chris-wood — CHANGES_REQUESTED** [review]
This looks to be in the right direction! I have some suggestions that aim to fix some editorial issues, reduce text, and simplify notation. We should note also that this assumes the VOPRF document will be extended to support metadata.

File: draft-ietf-privacypass-protocol.md (hunk 7 above)
Comment (claucece) [reply draft-ietf-privacypass-protocol.md]: Yeah.. I didn't know how to make it as long as wanted in the TLS language. Changed now ;)

File: draft-ietf-privacypass-architecture.md (hunk 1 above)
Comment (claucece) [reply draft-ietf-privacypass-architecture.md]: Changed ;)

File: draft-ietf-privacypass-architecture.md (hunk 2 above)
Comment (claucece) [reply draft-ietf-privacypass-architecture.md:481]: Changed.

File: draft-ietf-privacypass-protocol.md (hunk 3 above)
Comment (claucece) [reply draft-ietf-privacypass-protocol.md]: Done ;)

File: draft-ietf-privacypass-protocol.md (hunk 4 above)
Comment (claucece) [reply draft-ietf-privacypass-protocol.md]: Yes! changed everywhere.

**Pull request closed without merging**